setOnException not called in warp-tls

I see a difference in behavior between warp and warp-tls. The latter doesn't call my setOnException handler.

I've made a fully reproducible repo (run with `stack run`) where you can comment/uncomment to switch between http/https and see the difference: https://github.com/k-bx/servant-custommonad-logging/blob/master/waitls-onexception-bug/src/Main.hs

```
$ curl -i localhost:8000/api/error-out
```

HTTP Gives:

```
Allocating scarce resource
Cleaning up
Error: caught an exception: Erroring out
CallStack (from HasCallStack):
  error, called at src/Main.hs:33:12 in main:Main. Call stack: CallStack (from HasCallStack):
  onExceptionAct, called at src/Main.hs:37:29 in main:Main
```

Switchging to https:

```
$ curl --insecure -i https://localhost:8000/api/error-out
```

Gives no output from the app.

Sure, I can check it out on latest wai/warp/http2. Just tried -- getting this build error:

```
http2                 > Configuring http2-2.0.0...
http2                 > build
http2                 > Preprocessing library for http2-2.0.0..
http2                 > Building library for http2-2.0.0..
http2                 > [ 1 of 45] Compiling Imports
http2                 > [ 2 of 45] Compiling Network.HPACK.Builder
http2                 > [ 3 of 45] Compiling Network.HPACK.HeaderBlock.Integer
http2                 >
http2                 > /tmp/stack31729/http2-2.0.0/Network/HPACK/HeaderBlock/Integer.hs:40:11: error:
http2                 >     Not in scope: type constructor or class ‘WriteBuffer’
http2                 >    |
http2                 > 40 | encode :: WriteBuffer -> (Word8 -> Word8) -> Int -> Int -> IO ()
http2                 >    |           ^^^^^^^^^^^
http2                 >
http2                 > /tmp/stack31729/http2-2.0.0/Network/HPACK/HeaderBlock/Integer.hs:85:27: error:
http2                 >     Not in scope: type constructor or class ‘ReadBuffer’
http2                 >    |
http2                 > 85 | decode :: Int -> Word8 -> ReadBuffer -> IO Int
http2                 >    |                           ^^^^^^^^^^
http2                 >
```

You can reproduce by cloning this repo https://github.com/k-bx/servant-custommonad-logging and then doing `cd waitls-onexception-bug && stack build --fast`

--- a/warp/Network/Wai/Handler/Warp/HTTP2.hs
+++ b/warp/Network/Wai/Handler/Warp/HTTP2.hs
@@ -6,6 +6,7 @@
 
 module Network.Wai.Handler.Warp.HTTP2 (http2) where
 
+import qualified Data.IORef as I
 import qualified Control.Exception as E
 import qualified Network.HTTP2.Server as H2
 import Network.Socket (SockAddr)
@@ -37,21 +38,29 @@ http2 conn ii addr settings readN app =
 
     http2server h2req aux response = do
         req <- toWAIRequest h2req aux
-        ResponseReceived <- app req $ \rsp -> do
+        ref <- I.newIORef Nothing
+        eResponseReceived <- E.try $ app req $ \rsp -> do
             h2rsp <- fromResponse settings ii req rsp
             pps <- fromPushPromises ii req
-            ex <- E.try $ response h2rsp pps
-            case ex of
-              Right () -> do
-                  logResponse h2rsp req
-                  mapM_ (logPushPromise req) pps
-              Left  e@(E.SomeException _)
-                -- killed by the local worker manager
-                | Just E.ThreadKilled  <- E.fromException e -> return ()
-                -- killed by the local timeout manager
-                | Just T.TimeoutThread <- E.fromException e -> return ()
-                | otherwise -> S.settingsOnException settings (Just req) e
+            I.writeIORef ref $ Just (h2rsp, pps)
+            _ <- response h2rsp pps
             return ResponseReceived
+        case eResponseReceived of
+          Right ResponseReceived -> do
+              Just (h2rsp, pps) <- I.readIORef ref
+              logResponse h2rsp req
+              mapM_ (logPushPromise req) pps
+          Left e@(E.SomeException _)
+            -- killed by the local worker manager
+            | Just E.ThreadKilled  <- E.fromException e -> return ()
+            -- killed by the local timeout manager
+            | Just T.TimeoutThread <- E.fromException e -> return ()
+            | otherwise -> do
+                S.settingsOnException settings (Just req) e
+                let ersp = S.settingsOnExceptionResponse settings e
+                h2rsp' <- fromResponse settings ii req ersp
+                _ <- response h2rsp' []
+                logResponse h2rsp' req
         return ()
 
     toWAIRequest h2req aux = toRequest ii settings addr hdr bdylen bdy th

